### DISCUSSION TITLE
How to create RSS with article excerpt?

### DISCUSSION THREAD

**Discussion by jcubic** [facebook/docusaurus#9934] [Q&A] [answered]
I have this config:

```javascript
  presets: [
    [
      'classic',
      {
        blog: {
          showReadingTime: true,
          editUrl: 'https://github.com/jcubic/lips-website/tree/docusaurus/docs/',
          feedOptions: {
            type: 'rss',
            limit: 10,
            copyright: `Copyright © ${new Date().getFullYear()} Jakub T. Jankiewicz`,
            createFeedItems: ({ blogPosts, defaultCreateFeedItems, ...rest }) => {
              return defaultCreateFeedItems({
                blogPosts: blogPosts.map(({content, ...rest}) => {
                  const excerpt = content.replace(/<!--\s*truncate\s*-->[\s\S]*$/, '').trim();
                  console.log({ ...rest, content: excerpt });
                  return {
                    ...rest,
                    content: excerpt
                  }
                }),
                ...rest
              });
            }
          }
        },
```

The excerpt is correct, but RSS includes the whole article. Here is a repro:

https://codesandbox.io/p/devbox/blissful-bose-nxrvlq?file=%2Fdocusaurus.config.js%3A68%2C22

Unfortunately, you can't edit markdown files on Code Sandbox, they render as documents like README.

I see old discussion to make a way to add full blog article, but I don't see anywhere the option to turns this off: https://github.com/facebook/docusaurus/issues/3719.

**Comment by jcubic** [comment, marked as answer]
It's impossible, the blog plugin that create the feed just ignores the content and read HTML from the disk with cheerio [feed.ts](https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-plugin-content-blog/src/feed.ts), but there is a workaround. See comments.

**Reply by slorber** [reply]
It is possible, you can alter the result returned by defaultCreateFeedItem, or plainly create your own custom feed items.

You have total control over feed items

**Reply by ilg-ul** [reply]
It would be interesting to know how to do it.

I also tried and found out that the feed records contain large HTML text strings that, for me, were too difficult to parse to keep only the excerpt.

**Reply by jcubic** [reply]
@slorber, But there are no `<!--truncate-->` in output. How I'm supposed to create an excerpt?

**Reply by jcubic** [reply]
I've added a proposal to add an option for this.  https://docusaurus.io/feature-requests/p/allow-to-use-excerpt-of-blog-article-in-feed Since I don't think it's possible.

**Reply by jcubic** [reply]
@ilg-ul I will try to use Markdown parser and just ignore defaultCreateFeedItem (which is the only documented way to create feed), and maybe I will come up with something. Will share if I create something working.

**Reply by slorber** [reply]
The API gives you the ability to implement whatever you want, including creating your own custom marker, or hardcoding feed entries, or reading the feed data from another file or a CMS or git or anything.

https://stackblitz.com/edit/github-9prdtu?file=docusaurus.config.ts

```tsx
createFeedItems: async ({blogPosts,...params}) => {
  const feedItems = await params.defaultCreateFeedItems({blogPosts, ...params});
  feedItems.forEach((feedItem,index) => {
    const blogPost = blogPosts[index]!;
    feedItem.content = (blogPostst.metadata.frontMatter.custom_feed_content as string) 
      ?? blogPost.metadata.description 
      ?? blogPost.metadata.title;
  })
  return feedItems;
},
```

Docusaurus will add to the feed what you tell it to, and only provides a "default" opinionated behavior, but again you have 100% freedom to do whatever you want, including:


```tsx
createFeedItems: () => [{content: "my hardcoded blog post content"}]
```

Normally, creating feed items yourself should be relatively intuitive if you follow the TypeScript types you will know which attributes can be added to each feed entry.

We use the [feed](https://github.com/jpmonette/feed) library under the hood in case you prefer to read actual documentation.

> @slorber, But there are no `<!--truncate-->` in output. How I'm supposed to create an excerpt?


These markers are only used internally by our mdx loader to truncate posts for paginated list views. And comments are stripped from the HTML output as an optimization to reduce file sizes. 

If you want to truncate the feed, you can't necessarily reuse that marker. You could create another one and even have different excerps for RSS feeds and site pagination views.

---

Now, I'm open to implementing this feature as a "shortcut", but I'm not sure we should correlate this to the original truncate marker, and maybe add some kind of other marker such as `<div className="feed-truncate"/>` ?

**Reply by jcubic** [reply]
Thanks, this one works for me:

```typescript
import marked from 'marked';

createFeedItems: async ({ blogPosts,...params }) => {
    const feedItems = await params.defaultCreateFeedItems({ blogPosts, ...params });
    feedItems.forEach((feedItem,index) => {
        const blogPost = blogPosts[index]!;
        const excerpt = blogPost.content.replace(/<!--\s*truncate\s*-->[\s\S]*$/, '').trim();
        feedItem.content = marked.parse(excerpt) as string;
    });
    return feedItems;
}
```

`marked.parse` return type is `string | Promise<string>` but async only happen if you use plugins that are marked as async, so it's safe to cast to string. See [marked/discussions/3219](https://github.com/markedjs/marked/discussions/3219).

**Reply by slorber** [reply]
That would work but keep in mind Marked and MDX probably work very differently. 

If you add some MDX JSX, internal links or images with relative paths before the blog truncation marker, you will likely encounter problems in your feed.

Particularly with MDX, you cannot easily get the real output without actually rendering the content and then reading it from HTML.

See also https://github.com/facebook/docusaurus/issues/5664

**Reply by jcubic** [reply]
Sure, but in my case, I use normal Markdown files for blog.
